Advanced Helm chart customization doc (#11480)
istio/istio.io: content/en/docs/setup/additional-setup/customize-installation-helm/index.md

{{< text bash >}}
$ helm template istio/gateway --post-renderer ./kustomize.sh
{{< /text >}}

In the output, check for the newly added `sysctl` attribute for `ingress-gateway` pod:

{{< text yaml >}}
        - mountPath: /etc/envoy
          name: envoy-config
      securityContext:
        sysctls:
        - name: net.netfilter.nf_conntrack_tcp_timeout_close_wait
          value: "10"
{{< /text >}}

### Apply the patch using Helm Post Renderer

Use the below command to install an Istio ingress-gateway, applying our customization using Helm `post-renderer`:

{{< text bash >}}
$ kubectl create ns istio-ingress
$ helm upgrade -i istio-ingress istio/gateway --namespace istio-ingress --wait --post-renderer ./kustomize.sh
{{< /text >}}

Examine the ingress-gateway deployment, you will see the newly manipulated `sysctl` value:

{{< text yaml >}}
apiVersion: apps/v1
kind: Deployment
metadata:
  …
  name: istio-ingress
  namespace: istio-ingress
spec:
  template:
    metadata:
      …
    spec:
      securityContext:
        sysctls:
        - name: net.netfilter.nf_conntrack_tcp_timeout_close_wait
          value: "10"
{{< /text >}}

## References

> 1. [Gloo Edge - Last Mile Helm Chart Customization](https://docs.solo.io/gloo-edge/latest/installation/gateway/kubernetes/helm_advanced/)

Reviewer: I don't think we need to include this; users shouldn't need this with the doc we have

Thanks to the Solo team for the great doc though!

Author: Updated

---
title: Advanced Helm Chart Customization
description: Describes how to customize installation configuration options when installing with helm.
weight: 55
keywords: [profiles,install,helm]
owner: istio/wg-environments-maintainers
test: n/a
---

## Prerequisites

Before you begin, check the following prerequisites:

1. [Download the Istio release](/docs/setup/getting-started/#download).
1. Perform any necessary [platform-specific setup](/docs/setup/platform-setup/).
1. Check the [Requirements for Pods and Services](/docs/ops/deployment/requirements/).

> 1. [Usage of helm for Istio installation](/docs/setup/install/helm)

Author: Done

1. Helm version that supports post rendering. (>= 3.1)
1. kubectl or kustomize.

## Advanced Helm Chart Customization

Istio's helm chart tries to incorporate most of the attributes needed by users for their specific requirements. However, it does not
contain every possible Kubernetes value you may want to tweak. While it is not practical to have such a mechanism in place, in this
document we will demonstrate a method which would allow you to do some advanced helm chart customization without the need to directly
modify Istio's helm chart. Using Helm post rendering capability, you can tweak the installation manifests to meet your requirements easily.
Post rendering gives the flexibility to manipulate, configure, and/or validate rendered manifests before they are installed by Helm.
This enables users with advanced configuration needs to use tools like Kustomize to apply configuration changes without the need
for any additional support from the original chart maintainers.

> In this example, we will add a `sysctl` value to the Istio’s `ingress-gateway` deployment. We are going to:

Reviewer 2: Should be `to Istio's` or `to the Istio`

Author: Done

> 1. Create a `sysctl` deployment customization patch template.

Author: I have changed it, can you check if this looks better now?

### Create Deployment Customization

First, we create a `sysctl` patch file, adding a `securityContext` to the `ingress-gateway` pod with the additional attribute:

{{< text bash >}}
$ cat > sysctl-ingress-gw-customization.yaml <<EOF
apiVersion: apps/v1
kind: Deployment
metadata:
  name: ingress-gateway
spec:
  template:
    spec:
      securityContext:
          sysctls:
          - name: net.netfilter.nf_conntrack_tcp_timeout_close_wait
            value: "10"
EOF
{{< /text >}}

The below shell script helps to bridge the gap between Helm post renderer and Kustomize, as the former works with `stdin/stdout`
and latter works with files.

{{< text bash >}}
$ cat > kustomize.sh <<EOF
#!/bin/sh
cat > base.yaml
exec kubectl kustomize # you can also use "kustomize build ." if you have it installed.
EOF
chmod +x ./kustomize.sh
{{< /text >}}

Finally, let us create the `kustomization` yaml file, which is the input for `kustomize`
with the set of resources and associated customization details.

{{< text bash >}}
$ cat > kustomization.yaml <<EOF
resources:
- base.yaml
patchesStrategicMerge:
- sysctl-ingress-gw-customization.yaml
EOF
{{< /text >}}

> ## Apply and Verify the Customization

Reviewer 2: Kustomization file

Author: Changed

Now that the customization file is ready, let us use Helm to make sure this gets applied properly.

### Add the Helm repository for Istio

{{< text bash >}}
$ helm repo add istio https://istio-release.storage.googleapis.com/charts
$ helm repo update
{{< /text >}}

### Render and Verify using Helm Template

We can use Helm `post-renderer` to validate rendered manifests before they are installed by Helm

{{< text bash >}}
$ helm template istio/gateway --post-renderer ./kustomize.sh
{{< /text >}}

In the output, check for the newly added `sysctl` attribute for `ingress-gateway` pod:

{{< text yaml >}}
        - mountPath: /etc/envoy
          name: envoy-config
      securityContext:
        sysctls:
        - name: net.netfilter.nf_conntrack_tcp_timeout_close_wait
          value: "10"
{{< /text >}}

### Apply the patch using Helm Post Renderer

Use the below command to install an Istio ingress-gateway, applying our customization using Helm `post-renderer`:

{{< text bash >}}
$ kubectl create ns istio-ingress
$ helm upgrade -i istio-ingress istio/gateway --namespace istio-ingress --wait --post-renderer ./kustomize.sh
{{< /text >}}

> Examine the ingress-gateway deployment, you will see the newly manipulated `sysctl` value:

Reviewer 2: Should we have the describe command here to shows the output below?

---
title: Advanced Helm Chart Customization
description: Describes how to customize installation configuration options when installing with helm.
weight: 55
keywords: [profiles,install,helm]
owner: istio/wg-environments-maintainers
test: n/a
---

## Prerequisites

Before you begin, check the following prerequisites:

1. [Download the Istio release](/docs/setup/getting-started/#download).
1. Perform any necessary [platform-specific setup](/docs/setup/platform-setup/).
1. Check the [Requirements for Pods and Services](/docs/ops/deployment/requirements/).
1. [Usage of helm for Istio installation](/docs/setup/install/helm).
1. Helm version that supports post rendering. (>= 3.1)
1. kubectl or kustomize.

## Advanced Helm Chart Customization

Istio's helm chart tries to incorporate most of the attributes needed by users for their specific requirements. However, it does not
contain every possible Kubernetes value you may want to tweak. While it is not practical to have such a mechanism in place, in this
document we will demonstrate a method which would allow you to do some advanced helm chart customization without the need to directly

> modify Istio's helm chart. Using Helm post rendering capability, you can tweak the installation manifests to meet your requirements easily.

Reviewer 3: ```suggestion
modify Istio's helm chart.

### Using Helm with kustomize to post-render Istio charts

Using Helm post rendering capability, you can tweak the installation manifests to meet your requirements easily.
```

I find the paragraph to be too large. Maybe just break it up into 2 separate ones?

Post rendering gives the flexibility to manipulate, configure, and/or validate rendered manifests before they are installed by Helm.
This enables users with advanced configuration needs to use tools like Kustomize to apply configuration changes without the need
for any additional support from the original chart maintainers.

> In this example, we will add a `sysctl` value to Istio’s `ingress-gateway` deployment. We are going to:

Reviewer 3: ```suggestion
### Adding a value to an already existing chart

In this example, we will add a `sysctl` value to Istio’s `ingress-gateway` deployment. We are going to:
```

Author: Done

1. Create a `sysctl` deployment customization patch template.
1. Apply the patch using Helm post rendering.
1. Verify that the `sysctl` patch was correctly applied to the pods.

## Create the Kustomization

First, we create a `sysctl` patch file, adding a `securityContext` to the `ingress-gateway` pod with the additional attribute:

{{< text bash >}}
$ cat > sysctl-ingress-gw-customization.yaml <<EOF
apiVersion: apps/v1
kind: Deployment
metadata:
  name: ingress-gateway
spec:
  template:
    spec:
      securityContext:
          sysctls:
          - name: net.netfilter.nf_conntrack_tcp_timeout_close_wait
            value: "10"
EOF
{{< /text >}}

The below shell script helps to bridge the gap between Helm post renderer and Kustomize, as the former works with `stdin/stdout`
and latter works with files.

{{< text bash >}}
$ cat > kustomize.sh <<EOF
#!/bin/sh
cat > base.yaml
exec kubectl kustomize # you can also use "kustomize build ." if you have it installed.
EOF

> chmod +x ./kustomize.sh

Reviewer 3: ```suggestion
$ chmod +x ./kustomize.sh
```

{{< /text >}}

Finally, let us create the `kustomization` yaml file, which is the input for `kustomize`
with the set of resources and associated customization details.

{{< text bash >}}
$ cat > kustomization.yaml <<EOF
resources:
- base.yaml
patchesStrategicMerge:
- sysctl-ingress-gw-customization.yaml
EOF
{{< /text >}}

## Apply the Kustomization

Now that the Kustomization file is ready, let us use Helm to make sure this gets applied properly.

### Add the Helm repository for Istio

{{< text bash >}}
$ helm repo add istio https://istio-release.storage.googleapis.com/charts
$ helm repo update
{{< /text >}}

### Render and Verify using Helm Template

We can use Helm `post-renderer` to validate rendered manifests before they are installed by Helm

{{< text bash >}}

> $ helm template istio/gateway --post-renderer ./kustomize.sh | grep -B 4 -A 1 netfilter.nf_conntrack_tcp_timeout_close_wait

Reviewer 3: I'm getting this when I run this command

```bash

Error: error while running post render on files: error while running command /home/jacob/cluster/kind/kustomize.sh. error output:
error: no matches for Id Deployment.v1.apps/ingress-gateway.[noNs]; failed to find unique target for patch Deployment.v1.apps/ingress-gateway.[noNs]
: exit status 1
```

I think you want the following as you aren't using `nf_conntrack_tcp_timeout_close_wait`

```suggestion
$ helm template istio/gateway --post-renderer ./kustomize.sh
```

Reviewer 3: I'm still getting the same error when I looked at the updated page.

Reviewer 2: @jacob-delgado have you verified if this change now works for you?

Reviewer 3: It still doesn't work for me. I'm copying the commands from the webpage. 

```shell
jacob@jacob-build-3:~/faseela$ cat > sysctl-ingress-gw-customization.yaml <<EOF
> apiVersion: apps/v1
> kind: Deployment
> metadata:
>   name: istio-ingressgateway
> spec:
>   template:
>     spec:
>       securityContext:
>           sysctls:
>           - name: net.netfilter.nf_conntrack_tcp_timeout_close_wait
>             value: "10"
> EOF
jacob@jacob-build-3:~/faseela$
jacob@jacob-build-3:~/faseela$ cat > kustomize.sh <<EOF
> #!/bin/sh
> cat > base.yaml
> exec kubectl kustomize # you can also use "kustomize build ." if you have it installed.
> EOF
jacob@jacob-build-3:~/faseela$ chmod +x ./kustomize.sh
jacob@jacob-build-3:~/faseela$ cat > kustomization.yaml <<EOF
> resources:
> - base.yaml
> patchesStrategicMerge:
> - sysctl-ingress-gw-customization.yaml
> EOF
jacob@jacob-build-3:~/faseela$ helm repo add istio https://istio-release.storage.googleapis.com/charts
"istio" already exists with the same configuration, skipping
jacob@jacob-build-3:~/faseela$ helm repo update
Hang tight while we grab the latest from your chart repositories...
...Successfully got an update from the "istio" chart repository
...Successfully got an update from the "jetstack" chart repository
...Successfully got an update from the "gloo" chart repository
Update Complete. ⎈Happy Helming!⎈
jacob@jacob-build-3:~/faseela$ helm template istio/gateway --post-renderer ./kustomize.sh | grep -B 4 -A 1 netfilter.nf_conntrack_tcp_timeout_close_wait
Error: error while running post render on files: error while running command /home/jacob/faseela/kustomize.sh. error output:
error: no matches for Id Deployment.v1.apps/istio-ingressgateway.[noNs]; failed to find unique target for patch Deployment.v1.apps/istio-ingressgateway.[noNs]
: exit status 1

Use --debug flag to render out invalid YAML
```

Reviewer 3: When I do the same with the solo's guide at https://docs.solo.io/gloo-edge/latest/installation/gateway/kubernetes/helm_advanced/ it works just fine for me.

I am copying and pasting it with the webpage. I'm not typing this out manually. Maybe a copy/paste issue?

Reviewer 2: Thanks @jacob-delgado 

I just tried the same here and also see the error on this page. I did not try the solo.io page (Update: I did try the solo.io page and it worked).

Author: I will get back to this soon..on vacation currently.

Reviewer 2: OK. Comparing things, nothing jumps out as a cut/paste error. Maybe something related to the helm chart/repo.

Reviewer 3: Ok, I figured out the problem

```shell
$ helm template istio/gateway --post-renderer ./kustomize.sh
```

should be

```shell
$ helm template istio-ingressgateway istio/gateway --post-renderer ./kustomize.sh
```

Author: > Ok, I figured out the problem cc @kfaseela 
> 
> ```shell
> $ helm template istio/gateway --post-renderer ./kustomize.sh
> ```
> 
> should be
> 
> ```shell
> $ helm template istio-ingressgateway istio/gateway --post-renderer ./kustomize.sh
> ```

Thanks Jacob! I will try this out next week once am back from vacation and update the PR!

description: Describes how to customize installation configuration options when installing with helm.
weight: 55
keywords: [profiles,install,helm]
owner: istio/wg-environments-maintainers
test: n/a
---

## Prerequisites

Before you begin, check the following prerequisites:

1. [Download the Istio release](/docs/setup/getting-started/#download).
1. Perform any necessary [platform-specific setup](/docs/setup/platform-setup/).
1. Check the [Requirements for Pods and Services](/docs/ops/deployment/requirements/).
1. [Usage of helm for Istio installation](/docs/setup/install/helm).
1. Helm version that supports post rendering. (>= 3.1)
1. kubectl or kustomize.

## Advanced Helm Chart Customization

Istio's helm chart tries to incorporate most of the attributes needed by users for their specific requirements. However, it does not
contain every possible Kubernetes value you may want to tweak. While it is not practical to have such a mechanism in place, in this
document we will demonstrate a method which would allow you to do some advanced helm chart customization without the need to directly
modify Istio's helm chart.

### Using Helm with kustomize to post-render Istio charts

> Using Helm post rendering capability, you can tweak the installation manifests to meet your requirements easily.

Reviewer 2: I notice later we used `post-rendering`, so debating if we should use the back-ticks through the document.
```suggestion
Using the Helm `post-renderer`, you can tweak the installation manifests to meet your requirements easily.
```

> Post rendering gives the flexibility to manipulate, configure, and/or validate rendered manifests before they are installed by Helm.

Reviewer 2: ```suggestion
`Post-rendering` gives the flexibility to manipulate, configure, and/or validate rendered manifests before they are installed by Helm.
```

This enables users with advanced configuration needs to use tools like Kustomize to apply configuration changes without the need
for any additional support from the original chart maintainers.

### Adding a value to an already existing chart

In this example, we will add a `sysctl` value to Istio’s `ingress-gateway` deployment. We are going to:

1. Create a `sysctl` deployment customization patch template.

> 1. Apply the patch using Helm post rendering.

Reviewer 2: ```suggestion
1. Apply the patch using Helm `post-rendering`.
```

1. Verify that the `sysctl` patch was correctly applied to the pods.

## Create the Kustomization

First, we create a `sysctl` patch file, adding a `securityContext` to the `ingress-gateway` pod with the additional attribute:

{{< text bash >}}
$ cat > sysctl-ingress-gw-customization.yaml <<EOF
apiVersion: apps/v1
kind: Deployment
metadata:
  name: ingress-gateway
spec:
  template:
    spec:
      securityContext:
          sysctls:
          - name: net.netfilter.nf_conntrack_tcp_timeout_close_wait
            value: "10"
EOF
{{< /text >}}

> The below shell script helps to bridge the gap between Helm post renderer and Kustomize, as the former works with `stdin/stdout`

Reviewer 2: ```suggestion
The below shell script helps to bridge the gap between Helm `post-renderer` and Kustomize, as the former works with `stdin/stdout`
```

> and latter works with files.

Reviewer 2: ```suggestion
and the latter works with files.
```

with the set of resources and associated customization details.

{{< text bash >}}
$ cat > kustomization.yaml <<EOF
resources:
- base.yaml
patchesStrategicMerge:
- sysctl-ingress-gw-customization.yaml
EOF
{{< /text >}}

## Apply the Kustomization

Now that the Kustomization file is ready, let us use Helm to make sure this gets applied properly.

### Add the Helm repository for Istio

{{< text bash >}}
$ helm repo add istio https://istio-release.storage.googleapis.com/charts
$ helm repo update
{{< /text >}}

### Render and Verify using Helm Template

We can use Helm `post-renderer` to validate rendered manifests before they are installed by Helm

{{< text bash >}}
$ helm template istio/gateway --post-renderer ./kustomize.sh | grep -B 4 -A 1 netfilter.nf_conntrack_tcp_timeout_close_wait
{{< /text >}}

In the output, check for the newly added `sysctl` attribute for `ingress-gateway` pod:

{{< text yaml >}}
        - mountPath: /etc/envoy
          name: envoy-config
      securityContext:
        sysctls:
        - name: net.netfilter.nf_conntrack_tcp_timeout_close_wait
          value: "10"
{{< /text >}}

> ### Apply the patch using Helm Post Renderer

Reviewer 2: ```suggestion
### Apply the patch using Helm `Post-Renderer`
```

with the set of resources and associated customization details.

{{< text bash >}}
$ cat > kustomization.yaml <<EOF
resources:
- base.yaml
patchesStrategicMerge:
- sysctl-ingress-gw-customization.yaml
EOF
{{< /text >}}

## Apply the Kustomization

Now that the Kustomization file is ready, let us use Helm to make sure this gets applied properly.

### Add the Helm repository for Istio

{{< text bash >}}
$ helm repo add istio https://istio-release.storage.googleapis.com/charts
$ helm repo update
{{< /text >}}

### Render and Verify using Helm Template

We can use Helm `post-renderer` to validate rendered manifests before they are installed by Helm

{{< text bash >}}
$ helm template istio-ingress istio/gateway --namespace istio-ingress --post-renderer ./kustomize.sh | grep -B 4 -A 1 netfilter.nf_conntrack_tcp_timeout_close_wait
{{< /text >}}

In the output, check for the newly added `sysctl` attribute for `ingress-gateway` pod:

{{< text yaml >}}
        - mountPath: /etc/envoy
          name: envoy-config
      securityContext:
        sysctls:
        - name: net.netfilter.nf_conntrack_tcp_timeout_close_wait

>           value: "10"

Reviewer 2: Nit as it might vary, but my output here was (we are only told to check the sysctls):
```
          runAsNonRoot: true
          runAsUser: 1337
      securityContext:
        sysctls:
        - name: net.netfilter.nf_conntrack_tcp_timeout_close_wait
          value: "10"
```

{{< /text >}}

### Apply the patch using Helm `Post-Renderer`

Use the below command to install an Istio ingress-gateway, applying our customization using Helm `post-renderer`:

{{< text bash >}}
$ kubectl create ns istio-ingress
$ helm upgrade -i istio-ingress istio/gateway --namespace istio-ingress --wait --post-renderer ./kustomize.sh

> {{< /text >}}

Reviewer 2: Need to investigate more, but this is hanging:
```
istio-ingress        pod/istio-ingress-665695f7c8-zlk94                        0/1     SysctlForbidden   0          90s
istio-ingress        pod/istio-ingress-665695f7c8-zm7qk                        0/1     SysctlForbidden   0          58s
istio-ingress        pod/istio-ingress-665695f7c8-zmvc9                        0/1     SysctlForbidden   0          102s
istio-ingress        pod/istio-ingress-665695f7c8-zpbl5                        0/1     SysctlForbidden   0          8s
```
